src/main/scala/scala/tools/refactoring/implementations/MarkOccurrences.scala

```diff
@@ -15,21 +15,107 @@ import scala.tools.refactoring.util.SourceWithMarker.Movement
 import scala.tools.refactoring.common.TracingHelpers
 import scala.tools.refactoring.util.SourceHelpers
 import scala.tools.refactoring.util.SourceWithSelection
+import scala.tools.refactoring.util.Casts._
 
 trait MarkOccurrences extends common.Selections with analysis.Indexes with common.CompilerAccess with common.EnrichedTrees with common.InteractiveScalaCompiler {
   import global._
   import TracingHelpers.toCompactString
 
+  private def tryFindMissingSymbol(treeWithoutSymbol: Tree, root: Tree): Symbol = {
+    treeWithoutSymbol match {
+      case treeWithoutSymbol: RefTree =>
+        tryFindMissingSymbol(treeWithoutSymbol: RefTree, root: Tree)
+
+      case _ =>
+        NoSymbol
+    }
+  }
+
+  /*
+   * In some cases (for example when dealing with refined self types), the selected tree might not
+   * have a symbol. This method attempts to find it anyway.
+   */
+  private def tryFindMissingSymbol(treeWithoutSymbol: RefTree, root: Tree): Symbol = {
```

Comment:
Thanks for pointing me to the puzzle! I normally also prefer not using `return`, but all the returnless alternatives I tried where either not working properly, or looked very awkward. But maybe I should try harder... I will give it a second try!